pkg/gatewayserver/io/basicstationlns/correlations/correlations.go

```diff
@@ -0,0 +1,92 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package correlations
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// DownlinkInfo is the information associated with a particular downlink.
+type DownlinkInfo struct {
+	correlationIDs []string
+	txTime         time.Time
+}
+
+// DownlinkCorrelation is used to maintain tokens to correlate Downlink messages with their corresponding TxConfitrmations.
+type DownlinkCorrelation struct {
```

Comment:
Aren't we going a bit over the top with the concurrency here? We have only one goroutine doing downlink traffic (the one started in `handleTraffic`), and I think can also do the uplink part in the "main" goroutine for the gateway conn, right?